hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala

```diff
@@ -104,6 +104,21 @@ object LowerMatrixIR {
           entriesFieldName, entriesFieldName,
           child.typ.colKey)))
 
+    case MatrixRename(child, globalMap, colMap, rowMap, entryMap) =>
+      var t = lower(child).rename(rowMap, globalMap)
+
+      if (colMap.nonEmpty) {
+        val newColsType = TArray(child.typ.colType.rename(colMap))
+        t = t.mapGlobals('global.insertFields(colsField -> 'global (colsField).castRename(newColsType)))
+      }
+
+      if (entryMap.nonEmpty) {
+        val newEntriesType = child.typ.entryArrayType.copy(elementType = child.typ.entryType.rename(entryMap))
+        t = t.mapRows('row.insertFields(entriesField -> 'row (entriesField).castRename(newEntriesType)))
```

Comment:
You can just cast the entire row type here, right? like just combine this with the rowMap to create the new combined row/entries type and cast it?

Reply:
theoretically yes, but I'm a bit worried about how that will affect optimization.

Reply:
actually no, we couldn't rename key fields that way.

Reply:
ah, right. The first rename goes through TableRename instead of the current thing, which handles the keys.

My comment was more about being able to cast at the top level, which will be basically be a no-op, as opposed to an InsertFields, which will have some logic in it. Seems like we should just be able to create a new rowType for the resulting table that reflects the renamed entries and cast directly, no? And same for the cols in globalType? Would you still be worried about optimization in this case?

Reply:
yes, I had the same thought while writing the lowering and decided to use insertion.

I'm happy to adjust it to a no-op and change it back if we see troublesome IRs in the future